### PR TITLE
fix(memory): scope history.md/CLAUDE.md/data to client repo (#36)

### LINKED ISSUE #36: History MD only in Agents
## Problem

`history.md` in the **Agents** repository accumulates entries from **every** repository that uses Agents-Core, instead of each repo keeping its own history at its own root.

As a result, the project-level journal is no longer project-level — it mixes interactions from unrelated codebases and loses its value as per-repo memory.

## Expected behaviour

When Claude Code runs inside repository `X`, `log_interaction(...)` should append to `X/history.md`. Each repo that uses Agents-Core keeps its own, isolated journal.

## Actual behaviour

All `log_interaction` calls — regardless of the client's working directory — write to the **single** `history.md` located inside the Agents install directory (e.g. `~/repos/Agents/history.md`).

## Root cause

`history.md`'s location is derived from `REPO_ROOT`, and `REPO_ROOT` is hard-coded to the Agents install directory rather than the client's current working repo.

- `src/engine/config.py:7`
  ```python
  ENGINE_DIR = os.path.dirname(__file__)
  REPO_ROOT  = os.path.abspath(os.path.join(ENGINE_DIR, '../..'))
  ```
  This always resolves to the directory where the Agents framework source lives.

- `src/memory/config.py:18-19`
  ```python
  HISTORY_FILE        = os.path.join(REPO_ROOT, 'history.md')
  HISTORY_ARCHIVE_DIR = os.path.join(REPO_ROOT, 'history')
  ```
  Consequently, `HistoryWriter` always writes to `{Agents}/history.md`.

Because the `Agents-Core` MCP server is typically registered **globally** (one install, many client repositories), every Claude Code session — no matter which repo it was launched from — ends up writing into the same file.

## Proposed fix

Separate two distinct roots that are currently conflated:

1. **Install root** — where `src/`, `agents/`, `skills/`, `implants/` live. Derived from `__file__` (current behaviour, correct).
2. **Client repo root** — where per-repo artifacts (`history.md`, `CLAUDE.md` managed sections, `history/` archive) live. Must be resolved **per session / per call**, not from `__file__`.

Candidate strategies for resolving the client repo root (non-exclusive):

- Use the MCP server process's **current working directory** at startup — stdio MCP subprocesses typically inherit the client's cwd, so each Claude Code session naturally gets the correct root.
- Walk up from cwd to the nearest `.git` (or `.claude/`) marker.
- Accept an explicit `repo_root` argument in `route_and_load` / `log_interaction` / `describe_repo`, set by the client wrapper.
- Respect an environment variable (e.g. `AGENTS_CLIENT_REPO_ROOT`) that the client wrapper sets.

## Impact

- `HistoryWriter` / `HistoryReader` / `HistoryStore` — must point at the client repo's `history.md` and `history/` archive.
- `HistoryStore.data_dir` — the vector store for semantic recall should also live under the client repo's `data/memory/` (so semantic recall is per-repo, consistent with the journal).
- `describe_repo` / managed `CLAUDE.md` section — same consideration: the memory section must be written into the client repo's `CLAUDE.md`, not the Agents install's.
- Existing history must remain readable: either leave old entries in place, or provide a one-shot migration that scans `{Agents}/history.md` and fans entries out to their origin repos where identifiable.

## Scope

This issue covers the history split. Related concerns (describe_repo target, data/memory path) are in-scope if they share the same `REPO_ROOT` confusion; anything else should go to a follow-up issue.

### PR DESCRIPTION
**Pull request by WonderMr**
## Summary

- Split `REPO_ROOT` into **`INSTALL_ROOT`** (shipped content + indexes over shipped content: `agents/`, `skills/`, `implants/`, capabilities, router/skills/implants vector stores) and **`get_client_repo_root()`** (per-session, lazy, memoized — resolves via `AGENTS_CLIENT_REPO_ROOT` env → walk-up from cwd for `.git`/`CLAUDE.md` → cwd).
- Client-scoped paths (`history.md`, `history/` archive, managed `CLAUDE.md` section, `.describe_hash`, `history_store` vector index, `AGENTS_DEBUG` logs) now land in the repo Claude Code was launched from, not in the Agents install dir.
- `src/memory/config.py` exposes path names via PEP 562 `__getattr__` so `from src.memory.config import HISTORY_FILE` keeps working; deprecated `REPO_ROOT`/`DATA_DIR`/`DEBUG_LOG_DIR` aliases on `src/engine/config.py` preserve older imports.
- `describe_repo` / `write_repo_summary` sandboxes now clamp `repo_path` to the client root (was: install root).

Fixes #36.

## Design notes

- **Two install modes both work.** In per-repo-install mode, the cwd walk-up finds the client repo's `.git`; in shared-install mode (one Agents clone, many client repos), Claude Code's stdio subprocess inherits the workspace cwd and the walk-up still resolves correctly.
- **No env injection in `init_repo.sh`.** Injecting `AGENTS_CLIENT_REPO_ROOT` at registration time would pin every future session to whichever repo ran `init_repo.sh` — reintroducing the bug for shared-install users. The env var is supported as a manual/test override only.
- **No migration** of the existing `{install}/history.md`. Entries have no repo provenance; any migration would be heuristic. Old file stays readable via `HistoryReader(history_path=...)`.

## Test plan

- [x] `pytest tests/ --ignore=tests/test_language.py` → **199 passed, 0 regressions** (the 6 `test_language.py` failures are pre-existing `langdetect` import errors unrelated to this PR).
- [x] `tests/test_config_client_root.py` (13 tests) — env > walk-up > cwd precedence; `~` expansion; symlink realpath; `.git` file (worktree); `CLAUDE.md` marker; install-root invariance; alias mapping; cache reset.
- [x] `tests/test_per_repo_memory.py` (8 tests) — `HistoryWriter`, `HistoryStore`, `RepoDescriber` defaults respect `AGENTS_CLIENT_REPO_ROOT`; install data stays rooted at the install dir.
- [x] Manual end-to-end smoke: `AGENTS_CLIENT_REPO_ROOT=/tmp/fakerepo python -c 'from src.memory import config; print(config.HISTORY_FILE)'` → `/tmp/fakerepo/history.md`; unset → walks up to `.git`.

## Follow-ups (deliberate non-goals)

- [ ] Per-call `repo_path` override on `log_interaction` / `read_history` — one-liner if a multi-workspace-per-server scenario surfaces.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Moderate risk because it changes how filesystem roots are resolved and used for memory/log paths and tool sandboxing, which could redirect writes/reads across environments if the client-root resolution is wrong.
> 
> **Overview**
> This PR introduces a **two-root path model**: install-scoped assets/indexes stay under `INSTALL_ROOT`/`INSTALL_DATA_DIR`, while per-repo memory artifacts (e.g. `history.md`, managed `CLAUDE.md` section, `.describe_hash`, `history_store`, debug logs) are now resolved under a lazily-detected `get_client_repo_root()` (env override → walk-up for `.git`/`CLAUDE.md` → cwd).
> 
> It updates memory code (`src/memory/config.py`, `RepoDescriber`, `HistoryWriter`/`Reader`/`Store`) and debug/error logging to use client-scoped paths via PEP 562 lazy attributes, and adjusts `describe_repo`/`write_repo_summary` sandbox validation to clamp `repo_path` within the client root using a new `commonpath`-based `_is_within` helper.
> 
> Adds focused tests for client-root resolution, per-repo memory isolation defaults, and sandbox containment edge cases (including filesystem-root boundary regression).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8e95180ae25ea6eba6b7c535ac51701e7f4028f4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/engine/config.py [hunk 1]
```diff
@@ -1,25 +1,99 @@
...
+        return resolved
+
+    cwd = Path(os.getcwd())
+    marker = _find_marker_upwards(cwd)
+    if marker is not None:
+        logger.info("client-repo-root: walk-up marker -> %s", marker)
+        return str(marker)
+
+    fallback = str(cwd.resolve())
+    logger.info("client-repo-root: fallback cwd -> %s", fallback)
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/engine/config.py]: Agreed — dropped all three to `logger.debug` in 6e17dd3. The server's root logger is configured at INFO, so these fired on every startup. `AGENTS_DEBUG=1` still surfaces them when diagnosing.

File: tests/test_config_client_root.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        monkeypatch.setenv("AGENTS_CLIENT_REPO_ROOT", str(tmp_path))
+        # Client-scoped values shift...
+        engine_config._reset_client_repo_root_cache()
+        assert engine_config.get_client_repo_root() == os.path.realpath(str(tmp_path))
+        # ...but install-scoped values do not.
+        assert engine_config.INSTALL_ROOT.endswith("Agents")
+        assert engine_config.AGENTS_DIR.endswith("agents")
+        assert engine_config.SKILLS_DIR.endswith("skills")
+        assert engine_config.IMPLANTS_DIR.endswith("implants")
+        assert engine_config.INSTALL_DATA_DIR.endswith("data")
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_config_client_root.py]: Good catch — CI checkouts and forks often rename the folder. Replaced the `.endswith(...)` chain with structural assertions against `os.path.join(real_install_root, ...)` in 6e17dd3, so the test now checks that `AGENTS_DIR` / `SKILLS_DIR` / `IMPLANTS_DIR` / `INSTALL_DATA_DIR` are direct children of `INSTALL_ROOT`, regardless of folder name.

File: tests/test_per_repo_memory.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        result = writer.append_entry("intent", "action", "outcome")
+        assert result["status"] == "recorded"
+        assert (client_root / "history.md").exists()
+
+    def test_does_not_touch_install_root(self, client_root):
+        writer = HistoryWriter()
+        writer.append_entry("intent", "action", "outcome")
+        install_history = os.path.join(engine_config.INSTALL_ROOT, "history.md.test_sentinel")
+        # Creating the file is a side-effect we don't expect — assert the
+        # writer's path never points at the install root.
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_per_repo_memory.py]: Fixed in 6e17dd3 — the dangling `install_history` sentinel is gone. See the reply on the next comment for the full replacement (portable containment check via `os.path.commonpath`).

File: tests/test_per_repo_memory.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        assert not writer.history_path.startswith(engine_config.INSTALL_ROOT + "/history.md")
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_per_repo_memory.py]: Agreed — string-prefix checks don't survive Windows path separators or path normalization differences. Replaced in 6e17dd3 with:

```python
history_path = os.path.realpath(writer.history_path)
install_root = os.path.realpath(engine_config.INSTALL_ROOT)
assert os.path.commonpath([history_path, install_root]) != install_root
```

`commonpath` gives a portable containment test: if the history path were inside the install root, their common prefix would equal the install root.

File: src/engine/config.py [hunk 5]
```diff
@@ -1,25 +1,101 @@
...
+
+    cwd = Path(os.getcwd())
+    marker = _find_marker_upwards(cwd)
+    if marker is not None:
+        logger.debug("client-repo-root: walk-up marker -> %s", marker)
+        return str(marker)
+
+    fallback = str(cwd.resolve())
+    logger.debug("client-repo-root: fallback cwd -> %s", fallback)
+    return fallback
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/engine/config.py]: Good catch — this is a concrete hazard for long-running daemons started from ephemeral dirs. Fixed in 0feb39b:

```python
try:
    cwd = Path(os.getcwd())
except (FileNotFoundError, OSError) as err:
    logger.warning(
        "client-repo-root: cwd unavailable (%s); falling back to INSTALL_ROOT. "
        "Set AGENTS_CLIENT_REPO_ROOT to pin the per-session memory target.",
        err,
    )
    return INSTALL_ROOT
```

Wrapped both `os.getcwd()` and the final `cwd.resolve()` fallback. The warning is emitted at WARNING level (visible regardless of root log level) so the unexpected fallback is never silent. The env-var override still wins, so operators can pin the target and bypass the fallback entirely.

Added a regression test (`test_cwd_unavailable_falls_back_to_install_root`) that monkeypatches `os.getcwd` to raise `FileNotFoundError` and asserts both the resolved root equals `INSTALL_ROOT` and the warning is emitted. Full suite: 200 passed.

File: src/server.py [hunk 6]
```diff
@@ -707,17 +707,19 @@ async def describe_repo(
...
             resolved = os.path.realpath(repo_path)
-            repo_root_resolved = os.path.realpath(REPO_ROOT) + os.sep
-            if resolved != repo_root_resolved.rstrip(os.sep) and not resolved.startswith(repo_root_resolved):
+            client_root_resolved = os.path.realpath(client_root) + os.sep
+            if resolved != client_root_resolved.rstrip(os.sep) and not resolved.startswith(client_root_resolved):
                 payload = {
                     "status": "error",
-                    "error": f"repo_path must be within {REPO_ROOT}",
+                    "error": f"repo_path must be within {client_root}",
                 }
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/server.py:746]: Good find — when `client_root` resolves to `/`, the `realpath(...) + os.sep` math produced `"//"` and `rstrip(os.sep)` yielded `""`, rejecting every `repo_path`.

Fixed in 6ee1173 by extracting a `_is_within(candidate, boundary)` helper built on `os.path.commonpath`, matching the pattern in `src/utils/prompt_loader.py:67`. Applied to both `describe_repo` and `write_repo_summary`. Also catches `ValueError` from `commonpath` for Windows cross-drive paths. Regression covered by `tests/test_server_sandbox.py` (incl. a POSIX-root-boundary case).

File: src/server.py [hunk 7]
```diff
@@ -796,15 +798,16 @@ async def write_repo_summary(
...
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/server.py:833]: Addressed together with the matching `describe_repo` bug in 6ee1173 — both now delegate to the shared `_is_within` helper in `src/server.py`, which uses `os.path.commonpath` and catches `ValueError` for Windows cross-drive paths. No more hand-rolled separator math.

File: tests/test_config_client_root.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    engine_config._reset_client_repo_root_cache()
+    yield
+    engine_config._reset_client_repo_root_cache()
+
+
+class TestResolutionOrder:
+    def test_env_override_wins(self, tmp_path, monkeypatch):
+        monkeypatch.setenv("AGENTS_CLIENT_REPO_ROOT", str(tmp_path))
+        # Move cwd somewhere unrelated to prove env takes priority.
+        monkeypatch.chdir("/")
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_config_client_root.py]: Fixed in 6ee1173 — switched to a `tmp_path / "unrelated_cwd"` subdir that exists on every platform. The assertion (env override wins) is unchanged.

File: tests/test_config_client_root.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    def test_env_override_expands_tilde(self, monkeypatch):
+        monkeypatch.setenv("AGENTS_CLIENT_REPO_ROOT", "~")
+        resolved = engine_config.get_client_repo_root()
+        assert resolved == os.path.realpath(os.path.expanduser("~"))
+
+    def test_env_override_realpaths_symlink(self, tmp_path, monkeypatch):
+        real = tmp_path / "real"
+        real.mkdir()
+        link = tmp_path / "link"
+        link.symlink_to(real)
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_config_client_root.py]: Fixed in 6ee1173:

```python
try:
    link.symlink_to(real)
except OSError as exc:
    pytest.skip(f"symlink creation not supported here: {exc}")
```

Windows and some restricted CI sandboxes deny symlink creation; the skip keeps the suite portable while still exercising the realpath path on POSIX.

File: tests/test_per_repo_memory.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    monkeypatch.setenv("AGENTS_CLIENT_REPO_ROOT", str(root))
+    engine_config._reset_client_repo_root_cache()
+    yield root
+    engine_config._reset_client_repo_root_cache()
+
+
+class TestHistoryWriterDefault:
+    def test_writes_into_client_root(self, client_root):
+        writer = HistoryWriter()
+        assert writer.history_path == str(client_root / "history.md")
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_per_repo_memory.py:42]: Confirmed on macOS — `/var` → `/private/var` would make every `client_root / ...` comparison fail since the production resolver calls `os.path.realpath`.

Fixed in 8e95180 by having the `client_root` fixture yield `root.resolve()`, so every downstream assertion compares apples to apples. Matches the pattern already used in `tests/test_config_client_root.py`.

```python
@pytest.fixture
def client_root(tmp_path, monkeypatch):
    root = tmp_path / "client_repo"
    root.mkdir()
    monkeypatch.setenv("AGENTS_CLIENT_REPO_ROOT", str(root))
    engine_config._reset_client_repo_root_cache()
    yield root.resolve()  # collapses symlinked ancestors (macOS)
    engine_config._reset_client_repo_root_cache()
```

File: tests/test_server_sandbox.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
+"""Regression tests for the `_is_within` sandbox helper used by
+`describe_repo` / `write_repo_summary` (PR #37 review feedback).
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_server_sandbox.py]: Fixed in 8e95180 — the docstring now anchors on "issue #36" rather than self-referencing this PR. Cleaner when tracking the regression's origin later.

File: tests/test_server_sandbox.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        assert not _is_within(str(tmp_path), str(child))
+
+    def test_filesystem_root_as_boundary_accepts_descendants(self):
+        """Regression: the old ``"/" + os.sep`` math rejected every path
+        when the boundary was the filesystem root. ``commonpath`` handles
+        this cleanly."""
+        if os.name != "posix":
+            # On Windows, "/" isn't the true drive root; skip — the
+            # Windows-specific ``ValueError`` path is covered below.
+            return
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_server_sandbox.py]: Agreed — `return` reports as pass and the "covered below" note pointed at nothing. Fixed in 8e95180:

```python
if os.name != "posix":
    pytest.skip("filesystem-root semantics are POSIX-specific")
```

CI now reports skipped (not passed) on Windows. Also dropped the misleading comment about a Windows-specific test that doesn't exist in this file.
